### ISSUE TITLE
Turtle's `arg` does not support backtracking

### ISSUE THREAD

**Issue by Gabriel439** [Gabriel439/Haskell-Turtle-Library#119]
A `turtle` user reported this when they were attempting to use `Turtle.Options.arg` to simulate sub commands.  Here is the minimal reproducing example of the problem the user encountered:

```haskell
{-# LANGUAGE OverloadedStrings #-}

import qualified Data.Text as Text
import Turtle

example = argMatch "foo" empty <|> argMatch "bar" empty

main = do
    z <- options "" example
    print z

argMatch text =
    arg (\t -> if t == text then Just t else Nothing)
       (fromString (Text.unpack text))
```

This correctly parses the first alternative, but never parses the second alternative:

```bash
$ ./example foo
"foo"
$ ./example bar
Usage: options.hs (FOO | BAR)

Available options:
  -h,--help                Show this help text
```

The issue is that if an `arg`-based parser fails it does not allow further processing by other alternatives.  Here is another example using built-in `turtle` argument parsers:

```haskell
Usage: options.hs (FOO | BAR)
{-# LANGUAGE OverloadedStrings #-}

import Turtle

example = fmap (format d) (argInt "foo" empty) <|> argText "bar" empty

main = do
    z <- options "" example
    print z
```

A user might expect this to fall back on parsing a `Text` if `Int` parsing fails, but it does not:

```bash
$ ./example 5
"5"
$ ./example five
Usage: options.hs (FOO | BAR)

Available options:
  -h,--help                Show this help text
```

**Comment by Gabriel439** [comment]
Hmmmm.  It seems like this might be an upstream issue in `optparse-applicative`.  Both `arg` and `opt` are implemented in almost the exact same way, but `opt` backtracks correctly on parse failure.  Here is an example of the equivalent `opt` program succeeding:

```haskell
{-# LANGUAGE OverloadedStrings #-}

import qualified Data.Text as Text
import Turtle

example = optMatch "foo" 'f' empty <|> optMatch "bar" 'b' empty

main = do
    z <- options "" example
    print z

optMatch text c =
    opt (\t -> if t == text then Just t else Nothing)
       (fromString (Text.unpack text)) c
```

```bash
$ ./example --foo foo
"foo"
$ ./example --bar bar
"bar"
```

**Comment by Gabriel439** [comment]
I opened up an issue against the `optparse-applicative` library for this: https://github.com/pcapriotti/optparse-applicative/issues/166

**Comment by Gabriel439** [comment]
The linked ticket clarifies that `optparse-applicative` essentially parses at "two levels", one of which backtracks and one of which does not backtrack:

* First it parses the "skeleton" of the command line, meaning the numbers of arguments and any flags
* Then it parses the "payload" of the command line (i.e. the values that it actually extracts)

Parsing the "skeleton" backtracks, but parsing the "payload" does not.  A malformed payload is an unrecoverable parse error.

So the reason that `optparse-applicative` does not backtrack when parsing positional arguments is that their contents are part of the payload.  However, you can opt in to backtracking on a positional argument by using the `subcommand` functionality, which treats a positional argument as being part of the skeleton and not the payload.

So in other words the idiomatic solution to this problem is to add `subcommand` functionality to `turtle`.
